### PR TITLE
Add support to change PathSeperator

### PR DESCRIPTION
**Pull request by lukeb2e**
Hey,

I encountered the problem that my registry does not support arbitrary paths for images.

To counter this I added support to exchange "/" with any customizable seperator via cli.

I also had an issue with the login on my quay enterprise registry, hence why I needed to change api/v1/registry/client/client.go:114 to use repository instead of registry. 

Kind regards

### CODE REVIEW

**Comment by lukeb2e** [comment]
I don't know why the CI build keeps failing - might be that it is unable to access the DOCKERHUB_USERNAME & DOCKERHUB_PASSWORD.

`The command "docker login -u ${DOCKERHUB_USERNAME} -p ${DOCKERHUB_PASSWORD}" failed and exited with 125 during .`

**Comment by ivanilvescmp** [comment]
> I don't know why the CI build keeps failing - might be that it is unable to access the DOCKERHUB_USERNAME & DOCKERHUB_PASSWORD.
> 
> `The command "docker login -u ${DOCKERHUB_USERNAME} -p ${DOCKERHUB_PASSWORD}" failed and exited with 125 during .`

Thanks for this contribution. I'll be looking into it ASAP!

File: api/v1/registry/client/client.go [hunk 1]
```diff
@@ -113,11 +113,16 @@ func (cli *RegistryClient) Ping() error {
 func (cli *RegistryClient) Login(username, password string) error {
 	tk, err := auth.NewToken(cli.URL(), username, password, "registry:catalog:*")
 	if err != nil {
-		if username == "" && password == "" {
-			return nil
-		}
+		log.Debugf("Try to login with less permissions (repository:catalog:*)")
+		tk, err = auth.NewToken(cli.URL(), username, password, "repository:catalog:*")
```
Comment (ivanilves) [inline api/v1/registry/client/client.go]: good move 👌

File: api/v1/v1.go [hunk 2]
```diff
@@ -5,12 +5,13 @@ package v1
 import (
 	"bufio"
 	"fmt"
-	log "github.com/sirupsen/logrus"
 	"io"
 	"runtime"
 	"strings"
 	"time"
 
+	log "github.com/sirupsen/logrus"
```
Comment (ivanilves) [inline api/v1/v1.go]: makes sense 👌

File: api/v1/v1.go [hunk 3]
```diff
@@ -49,6 +50,8 @@ type PushConfig struct {
 	Registry string
 	// UpdateChanged tells us if we will re-push (update/overwrite) images having same tag, but different digest
 	UpdateChanged bool
+	// PathSeperator defines which path seperator to use (default: "/")
```
Comment (ivanilves) [inline api/v1/v1.go]: `PathSeperator` ? 🤔 Maybe `PathSeparator` ?

File: main.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ type Options struct {
 	PushRegistry       string        `short:"r" long:"push-registry" description:"[Re]Push pulled images to a specified remote registry" env:"PUSH_REGISTRY"`
 	PushPrefix         string        `short:"R" long:"push-prefix" description:"[Re]Push pulled images with a specified repo path prefix" env:"PUSH_PREFIX"`
 	PushUpdate         bool          `short:"U" long:"push-update" description:"Update our pushed images if remote image digest changes" env:"PUSH_UPDATE"`
+	PathSeperator      string        `short:"s" long:"path-seperator" default:"/" description:"Configure path seperator for registries that only allow single folder depth" env:"PATH_SEPERATOR"`
```
Comment (ivanilves) [inline main.go]: > `PathSeperator` ? 🤔 Maybe `PathSeparator` ?

File: main.go [hunk 5]
```diff
@@ -162,6 +163,7 @@ func main() {
 				Registry:      o.PushRegistry,
 				Prefix:        o.PushPrefix,
 				UpdateChanged: o.PushUpdate,
+				PathSeperator: o.PathSeperator,
```
Comment (ivanilves) [inline main.go]: > `PathSeperator` ? 🤔 Maybe `PathSeparator` ?

File: repository/repository.go [hunk 6]
```diff
@@ -89,6 +89,15 @@ func (r *Repository) Path() string {
 	return path
 }
 
+// PushPath gives us sanitized repository path and checks if subdirectories are allowed
```
Comment (ivanilves) [inline repository/repository.go]: This functions does not actually check anything 🤔 Maybe

> // PushPath gives us repository path having custom path element separator

... or something like this. 

WDYT?

File: main.go (hunk 5 above)
Comment (lukeb2e) [reply main.go]: haha true, that is a mix up

File: api/v1/v1.go (hunk 3 above)
Comment (lukeb2e) [reply api/v1/v1.go]: Changed it to `PathSeparator`

**Comment by lukeb2e** [comment]
One shouldn't rebase before the first coffee...

**Comment by ivanilves** [comment]
Good job @lukeb2e 

I'm merging this. We have some errors in CI, but they may be sporadic or not important.

**Pull request merged**